Frameworks/CoreGraphics/CGPath.mm

```diff
@@ -283,10 +283,64 @@ CFTypeID CGPathGetTypeID() {
     return __CGPath::GetTypeID();
 }
 
+// A helper for determining the number of points per path element type
+static inline int CGPathGetExpectedPointCountForType(CGPathElementType type) {
+    return type == kCGPathElementMoveToPoint || type == kCGPathElementAddLineToPoint ?
+               1 :
+               type == kCGPathElementAddQuadCurveToPoint ? 2 : type == kCGPathElementAddCurveToPoint ? 3 : 0;
+}
+
+// Create a mimic of CGPathElement that holds the points in a vector with a convenient copy constructor.
+struct _CGPathElementVector {
+    CGPathElementType _type;
+    std::vector<CGPoint> _points;
+    _CGPathElementVector(const CGPathElement el)
+        : _type(el.type), _points(el.points, el.points + CGPathGetExpectedPointCountForType(el.type)) {
+    }
+};
+
+// A struct to pass to the equality matching CGPathApply since only a single void* may be passed.
+struct CGPathElementMatch {
+    std::vector<_CGPathElementVector> elements;
+    bool equal = true;
+    int positionToMatch = 0;
+};
+
+// A function to pass to CGPathApply to determine whether two paths are equal.
+static void _CGPathApplyCheckEquality(void* pathElements, const CGPathElement* element1) {
+    if (!((CGPathElementMatch*)pathElements)->equal) {
+        return;
+    }
+    int i = ((CGPathElementMatch*)pathElements)->positionToMatch;
+    _CGPathElementVector element2 = ((CGPathElementMatch*)pathElements)->elements[i];
+    if (element2._type != element1->type) {
+        ((CGPathElementMatch*)pathElements)->equal = false;
+    } else {
+        for (int i = 0; i < CGPathGetExpectedPointCountForType(element1->type); i++) {
+            if (element1->points[i] != element2._points[i]) {
+                ((CGPathElementMatch*)pathElements)->equal = false;
+            }
+        }
+    }
+
+    ((CGPathElementMatch*)pathElements)->positionToMatch++;
+}
+
+// A function to pass to CGPathApply to retrieve the individual path elements to check equality against.
+static void _CGPathApplyGetElements(void* pathElements, const CGPathElement* element) {
+    _CGPathElementVector newElement = _CGPathElementVector(*element);
+    ((std::vector<_CGPathElementVector>*)pathElements)->emplace_back(newElement);
+}
+
 static Boolean __CGPathEqual(CFTypeRef cf1, CFTypeRef cf2) {
     if (!cf1 && !cf2) {
         return true;
     }
+
+    if (cf1 == cf2) {
```

Comment:
The above check is redundant with this one. Remove the above one in favour of this one